internal/kms/kms.go

```diff
@@ -269,6 +273,245 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
+// Options are ignored.
+func (k *Kms) ListDataKeyVersionDestructionJobs(ctx context.Context, scopeId string, _ ...Option) ([]*DataKeyVersionDestructionJobProgress, error) {
+	var jobs []*DataKeyVersionDestructionJobProgress
+	if err := k.reader.SearchWhere(ctx, &jobs, "scope_id=?", []interface{}{scopeId}); err != nil {
+		return nil, err
+	}
+	return jobs, nil
+}
+
+// MonitorTableRewrappingRuns checks for pending rewrapping runs for the
+// specified table name, and attempts to become the running run and
+// start rewrapping data in the specified table. This may be a long running operation.
+// Options are ignored.
+func (k *Kms) MonitorTableRewrappingRuns(ctx context.Context, tableName string, _ ...Option) (retErr error) {
+	const op = "kms.(Kms).MonitorTableRewrappingRuns"
+
+	rewrapFn, ok := tableNameToRewrapFn[tableName]
+	if !ok {
+		return errors.E(ctx, errors.WithMsg("no rewrapper for table %q", tableName), errors.WithOp(op))
+	}
+
+	run := allocDataKeyVersionDestructionJobRun()
+	err := k.reader.LookupWhere(ctx, &run, "is_running=true and table_name!=?", []interface{}{tableName})
+	switch {
+	case err == nil:
+		// Another run was already running
+		return nil
+	case errors.Match(errors.T(errors.RecordNotFound), err):
+		// Great, no running run, lets continue
+	default:
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to find any running runs"))
+	}
+	// Find the oldest or currently running run for our tablename
+	run = allocDataKeyVersionDestructionJobRun()
+	rows, err := k.reader.Query(ctx, oldestPendingOrRunningRun, []interface{}{tableName})
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to query pending runs for %q", tableName))
+	}
+	defer rows.Close()
+	for rows.Next() {
+		if err := k.reader.ScanRows(ctx, rows, &run); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan pending run for %q", tableName))
+		}
+	}
+	if run.KeyId == "" {
+		// No queued runs, lets try again later
+		return nil
+	}
+	if !run.IsRunning {
+		// No running job, lets try to become the running job
+		run.IsRunning = true
+		if _, err := k.writer.Update(ctx, &run, []string{"IsRunning"}, nil); err != nil {
+			// Unique error means some other run started running since we last checked.
+			if errors.Match(errors.T(errors.NotUnique), err) {
+				return nil
+			}
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to become running run for %q", tableName))
+		}
+	}
+
+	defer func() {
+		if retErr != nil {
+			// Create new context in case we failed because the context was canceled
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+		}
+		// Update the progress of this run
+		args := []interface{}{
+			sql.Named("key_id", run.KeyId),
+			sql.Named("table_name", tableName),
+		}
+		if _, err := k.writer.Exec(ctx, fmt.Sprintf(updateCompletedCountQueryTemplate, tableName), args); err != nil {
```

Comment:
so this query will updated the completed count to an accurate figure even if a previous run was killed and unable to update the completed count. however, this will leave an inaccurate completed count throughout the duration of this run's execution. I assume that's known + expected + and we don't really care since we only update after a run finishes anyway?

Reply:
Indeed, we don't do any mid-run updates, only at the end of the run. Since we break up the rewrapping per-table I think that's still okay as the user will get updates every time a table finishes. We could implement mid-run updates in a future update.